File: classes04/python/warpTriangle.py
import cv2
import numpy as np

def warpTriangle(img1, img2, tri1, tri2):
    """
    变换三角形
    """
    # 先求出三角形的包围盒
    r1 = cv2.boundingRect(tri1)
    r2 = cv2.boundingRect(tri2)

    print("tri1:{}".format(tri1))
    print("r1:{}".format(r1))
    print("tri2:{}".format(tri2))
    print("r2:{}".format(r2))

    # 使用包围盒剪裁图形
    img1Cropped = img1[r1[1]:r1[1]+r1[3],r1[0]:r1[0]+r1[2]]

    # 调整坐标系
    tri1Cropped = []
    tri2Cropped = []
    for i in range(0,3):
        tri1Cropped.append(((tri1[0][i][0] - r1[0]),(tri1[0][i][1] - r1[1])))
        tri2Cropped.append(((tri2[0][i][0] - r1[0]),(tri2[0][i][1] - r2[1])))
    
    # 获取变换矩阵
    warpMat = cv2.getAffineTransform(np.float32(tri1Cropped), np.float32(tri2Cropped))

    print("warpMat:{}".format(warpMat))

    # 进行仿射变换
    img2Cropped = cv2.warpAffine(img1Cropped,warpMat,(r2[2],r2[3]),None,flags=cv2.INTER_LINEAR,borderMode=cv2.BORDER_REFLECT_101)

    # 进行alpha混合
    mask = np.zeros((r2[3],r2[2],3), dtype=np.float32)
    cv2.fillConvexPoly(mask, np.int32(tri2Cropped),(1.0,1.0,1.0), 16, 0)

    img2Cropped = img2Cropped * mask

    img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] = img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] * ( (1.0, 1.0, 1.0) - mask )
    img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] = img2[r2[1]:r2[1]+r2[3], r2[0]:r2[0]+r2[2]] + img2Cropped

if __name__ == '__main__' :
    imgIn = cv2.imread("../data/images/kingfisher.jpg")
    imgOut = 255 * np.ones(imgIn.shape, dtype = imgIn.dtype)

    triIn = np.float32([[[360,50], [60,100], [300,400]]])
    triOut = np.float32([[[400,200], [160,270], [400,400]]])
    warpTriangle(imgIn, imgOut, triIn, triOut)

    color = (255, 150, 0)
    cv2.polylines(imgIn, triIn.astype(int), True, color, 2, cv2.LINE_AA)
    cv2.polylines(imgOut, triOut.astype(int), True, color, 2, cv2.LINE_AA)
  
    cv2.namedWindow("Input")
    cv2.imshow("Input", imgIn)

    cv2.namedWindow("Output")
    cv2.imshow("Output", imgOut)

    cv2.waitKey(0)

File: classes04/cpp/warpTriangle.cpp
#include <opencv2/opencv.hpp>
#include <stdlib.h>

using namespace cv;
using namespace std;

/**
 * @brief 
 * 
 * @param img1 
 * @param img2 
 * @param tri1 
 * @param tri2 
 */
void warpTriangle(Mat &img1, Mat &img2, vector<Point2f> tri1, vector<Point2f> tri2);

int main(int argc, char** argv)
{
    // 读取图片
    Mat imgInput = imread("../data/images/kingfisher.jpg", IMREAD_COLOR);

    imgInput.convertTo(imgInput, CV_32FC3, 1/255.0);

    Mat imgOut = Mat::ones(imgInput.size(), imgInput.type());
    imgOut = Scalar(1.0,1.0,1.0);

    // 输入三角形
    vector<Point2f> triIn;
    triIn.push_back(Point2f(360,50));
    triIn.push_back(Point2f(60,10));
    triIn.push_back(Point2f(300,400));

    // 输出三角形
    vector<Point2f> triOut;
    triOut.push_back(Point2f(400,200));
    triOut.push_back(Point2f(160,270));
    triOut.push_back(Point2f(400,400));

    warpTriangle(imgInput, imgOut, triIn, triOut);

    // 绘制CV_AA抗锯齿仅支持CV_8U3图片
    imgInput.convertTo(imgInput, CV_8UC3, 255.0);
    imgOut.convertTo(imgOut, CV_8UC3, 255.0);

    // 绘制线条用的颜色
    Scalar color = Scalar(255, 150, 0);

    // cv::polylines需要Point而不是Point2f
    vector <Point> triInInt, triOutInt;
    for(int i=0; i<3; i++)
    {
        triInInt.push_back(Point(triIn[i].x, triIn[i].y));
        triOutInt.push_back(Point(triOut[i].x, triOut[i].y));
    }

    int lineWidth = 2;
    polylines(imgInput, triInInt, true, color, lineWidth, CV_AA);
    polylines(imgOut, triOutInt, true, color, lineWidth, CV_AA);

    imshow("Input", imgInput);
    imshow("Output", imgOut);

    waitKey(0);

    return EXIT_SUCCESS;
}

void warpTriangle(Mat &img1, Mat &img2, vector<Point2f> tri1, vector<Point2f> tri2)
{
    // 获取包围盒
    Rect r1 = boundingRect(tri1);
    Rect r2 = boundingRect(tri2);

    // 截取图像
    Mat img1Cropped;
    img1(r1).copyTo(img1Cropped);

    imshow("img1Cropped", img1Cropped);

    vector<Point2f> tri1Cropped, tri2Cropped;
    vector<Point> tri2CroppedInt;
    for(int i=0; i<3; i++)
    {
        tri1Cropped.push_back(Point2f(tri1[i].x - r1.x, tri1[i].y - r1.y));
        tri2Cropped.push_back(Point2f(tri2[i].x - r2.x, tri2[i].y - r2.y));

        tri2CroppedInt.push_back(Point((int)(tri2[i].x - r2.x),(int)(tri2[i].y - r2.y)));
    }

    // 获取仿射变换矩阵
    Mat warpMat = getAffineTransform(tri1Cropped, tri2Cropped);

    // 应用仿射变换
    Mat img2Cropped = Mat::zeros(r2.height, r2.width, img1Cropped.type());
    warpAffine(img1Cropped, img2Cropped, warpMat, img2Cropped.size(), INTER_LINEAR, BORDER_REFLECT_101);

    imshow("warpAffine", img2Cropped);

    // 剔除掉三角形以外的像素
    Mat mask = Mat::zeros(r2.height, r2.width, CV_32FC3);
    fillConvexPoly(mask, tri2CroppedInt, Scalar(1.0, 1.0, 1.0), 16, 0);

    imshow("mask", mask);

    multiply(img2Cropped, mask, img2Cropped);
    imshow("mul1", img2Cropped);
    multiply(img2(r2), Scalar(1.0, 1.0, 1.0)-mask, img2(r2));
    imshow("mul2", img2);

    img2(r2) += img2Cropped;
}